File: Ent_UsAdmin/EmpresasPermitidas.vb
Imports System.ComponentModel
Imports System.Data.OleDb
Imports System.Data.SqlClient

Public Class EmpresasPermitidas
    Dim ncargando As Cargando
    Public idUsuario As String
    Private ButtonIndex = 0
    Private DragDropCursor As Cursor
    Private Sub EmpresasPermitidas_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        CheckForIllegalCrossThreadCalls = False
        ButtonIndex += 1

        ncargando = New Cargando
        ncargando.MonoFlat_Label1.Text = "Cargando Información"
        ncargando.Show()

        FlowEmpresasRestringidas.AllowDrop = True
        FlowEmpresasPermitidas.AllowDrop = True
        BackgroundWorker1.RunWorkerAsync()


    End Sub

    Private Sub BackgroundWorker1_DoWork(sender As Object, e As System.ComponentModel.DoWorkEventArgs) Handles BackgroundWorker1.DoWork
        iniciarconexionR()
        Dim comandoEmpresasPermitidas As OleDbCommand

        Dim consultaEmpresasPermitidas As String
        Dim readerEmpresasPermitidas As OleDbDataReader
        consultaEmpresasPermitidas = "select id,nombre from empresas where id in (select idempresa from empresaspermitidas where idusuario = '" & idUsuario & "')"
        comandoEmpresasPermitidas = New OleDbCommand
        comandoEmpresasPermitidas.Connection = conexionempresaR
        comandoEmpresasPermitidas.CommandText = consultaEmpresasPermitidas
        readerEmpresasPermitidas = comandoEmpresasPermitidas.ExecuteReader
        If readerEmpresasPermitidas.HasRows Then

            While readerEmpresasPermitidas.Read
                Dim botonempresa As New Bunifu.Framework.UI.BunifuFlatButton

                botonempresa.Normalcolor = Color.FromArgb(48, 55, 76)
                botonempresa.Iconimage = Nothing

                botonempresa.Size = New Size(197, 48)
                botonempresa.Name = readerEmpresasPermitidas("id").ToString
                botonempresa.Text = readerEmpresasPermitidas("nombre").ToString
                '  botonempresa.Tag = milector("ip")
                'AddHandler botonempresa.Click, AddressOf clickevenntAsync
                AddHandler botonempresa.MouseDown, AddressOf OnButtonMouseDown
                Me.Invoke(Sub()
                              FlowEmpresasPermitidas.Controls.Add(botonempresa)
                          End Sub)

            End While

        End If

        Dim comandoEmpresasRestringidas As OleDbCommand
        Dim consultaEmpresaRestringidas As String
        Dim readerEmpresasRestringidas As OleDbDataReader
        consultaEmpresaRestringidas = "select id,nombre from empresas where id not in (select idempresa from empresaspermitidas where idusuario = '" & idUsuario & "')"
        comandoEmpresasRestringidas = New OleDbCommand
        comandoEmpresasRestringidas.Connection = conexionempresaR
        comandoEmpresasRestringidas.CommandText = consultaEmpresaRestringidas
        readerEmpresasRestringidas = comandoEmpresasRestringidas.ExecuteReader
        If readerEmpresasRestringidas.HasRows Then

            While readerEmpresasRestringidas.Read
                Dim botonempresa As New Bunifu.Framework.UI.BunifuFlatButton

                botonempresa.Normalcolor = Color.FromArgb(48, 55, 76)
                botonempresa.Iconimage = Nothing

                botonempresa.Size = New Size(197, 48)
                botonempresa.Name = readerEmpresasRestringidas("id").ToString
                botonempresa.Text = readerEmpresasRestringidas("nombre").ToString
                '  botonempresa.Tag = milector("ip")
                'AddHandler botonempresa.Click, AddressOf clickevenntAsync
                AddHandler botonempresa.MouseDown, AddressOf OnButtonMouseDown
                Me.Invoke(Sub()
                              FlowEmpresasRestringidas.Controls.Add(botonempresa)
                          End Sub)

            End While

        End If




    End Sub

    Private Sub OnButtonMouseDown(sender As Object, e As MouseEventArgs)

        Dim button = DirectCast(sender, Bunifu.Framework.UI.BunifuFlatButton)

        Using bmp As New Bitmap(button.Width, button.Height)
            button.DrawToBitmap(bmp, New Rectangle(Point.Empty, button.Size))
            DragDropCursor = New Cursor(bmp.GetHicon)
        End Using

        button.Parent.DoDragDrop(button, DragDropEffects.Move)

    End Sub

    Private Sub OnFlowLayoutPanelDragEnter(sender As Object, e As DragEventArgs) _
        Handles FlowEmpresasPermitidas.DragEnter, FlowEmpresasRestringidas.DragEnter

        If e.AllowedEffect = DragDropEffects.Move AndAlso e.Data.GetDataPresent(GetType(Bunifu.Framework.UI.BunifuFlatButton)) Then
            e.Effect = DragDropEffects.Move
        End If

    End Sub

    Private Sub OnFlowLayoutPanelDragDrop(sender As Object, e As DragEventArgs) _
        Handles FlowEmpresasPermitidas.DragDrop, FlowEmpresasRestringidas.DragDrop

        Dim button = DirectCast(e.Data.GetData(GetType(Bunifu.Framework.UI.BunifuFlatButton)), Bunifu.Framework.UI.BunifuFlatButton)
        Dim destPanel = DirectCast(sender, FlowLayoutPanel)
        destPanel.Controls.Add(button)
        Dim targetLoc = destPanel.GetChildAtPoint(destPanel.PointToClient(New Point(e.X, e.Y)))

        If targetLoc IsNot Nothing Then
            Dim idx = destPanel.Controls.GetChildIndex(targetLoc)
            destPanel.Controls.SetChildIndex(button, idx)
        End If

    End Sub

    Private Sub OnFlowLayoutPanelGiveFeedback(sender As Object, e As GiveFeedbackEventArgs) _
        Handles FlowEmpresasPermitidas.GiveFeedback, FlowEmpresasRestringidas.GiveFeedback

        e.UseDefaultCursors = False
        Cursor.Current = DragDropCursor

    End Sub

    Private Sub EmpresasPermitidas_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        Dim num_controles As Integer = FlowEmpresasPermitidas.Controls.Count - 1
        For n As Integer = num_controles To 0 Step -1
            Dim ctrl As Control = FlowEmpresasPermitidas.Controls(n)
            FlowEmpresasPermitidas.Controls.Remove(ctrl)
            ctrl.Dispose()
        Next

        Dim num_controles2 As Integer = FlowEmpresasRestringidas.Controls.Count - 1
        For n As Integer = num_controles2 To 0 Step -1
            Dim ctrl As Control = FlowEmpresasRestringidas.Controls(n)
            FlowEmpresasRestringidas.Controls.Remove(ctrl)
            ctrl.Dispose()
        Next

    End Sub

    Private Sub BackgroundAplicar_DoWork(sender As Object, e As System.ComponentModel.DoWorkEventArgs) Handles BackgroundAplicar.DoWork
        iniciarconexionR()

        For Each ctrl As Control In FlowEmpresasPermitidas.Controls
            If TypeOf (ctrl) Is Bunifu.Framework.UI.BunifuFlatButton Then
                Dim btnctrl As Bunifu.Framework.UI.BunifuFlatButton = ctrl
                Dim comandoEmpresaPermitida As OleDbCommand
                Dim consultaEmpresaPermitida As String
                consultaEmpresaPermitida = "if not exists(select id from empresaspermitidas where idempresa = '" & btnctrl.Name & "' and idusuario = '" & idUsuario & "')
                                            begin
                                              insert into empresasPermitidas values('" & btnctrl.Name & "','" & idUsuario & "')
                                               end"
                comandoEmpresaPermitida = New OleDbCommand
                comandoEmpresaPermitida.Connection = conexionempresaR
                comandoEmpresaPermitida.CommandText = consultaEmpresaPermitida
                comandoEmpresaPermitida.ExecuteNonQuery()



            End If
        Next


        For Each ctrl As Control In FlowEmpresasRestringidas.Controls
            If TypeOf (ctrl) Is Bunifu.Framework.UI.BunifuFlatButton Then
                Dim btnctrl As Bunifu.Framework.UI.BunifuFlatButton = ctrl
                Dim comandoEmpresaRestringida As OleDbCommand
                Dim consultaEmpresaRestringida As String
                consultaEmpresaRestringida = "if exists(select id from empresaspermitidas where idempresa = '" & btnctrl.Name & "' and idusuario = '" & idUsuario & "')
                                             begin
                                            delete from empresasPermitidas where idempresa = '" & btnctrl.Name & "' and idusuario = '" & idUsuario & "' 
                                            end"
                comandoEmpresaRestringida = New OleDbCommand
                comandoEmpresaRestringida.Connection = conexionempresaR
                comandoEmpresaRestringida.CommandText = consultaEmpresaRestringida
                comandoEmpresaRestringida.ExecuteNonQuery()



            End If
        Next

    End Sub

    Private Sub BackgroundWorker1_RunWorkerCompleted(sender As Object, e As RunWorkerCompletedEventArgs) Handles BackgroundWorker1.RunWorkerCompleted
        ncargando.Close()

    End Sub

    Private Sub BackgroundAplicar_RunWorkerCompleted(sender As Object, e As RunWorkerCompletedEventArgs) Handles BackgroundAplicar.RunWorkerCompleted
        ncargando.Close()
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        ncargando = New Cargando
        ncargando.MonoFlat_Label1.Text = "Aplicando cambios"
        ncargando.Show()

        BackgroundAplicar.RunWorkerAsync()

    End Sub

    Private Sub Panel1_Paint(sender As Object, e As PaintEventArgs) Handles Panel1.Paint

    End Sub

    Private Sub Panel1_MouseDown(sender As Object, e As MouseEventArgs) Handles Panel1.MouseDown
        If e.Button = MouseButtons.Left Then
            MoveForm(Me)
        End If
    End Sub
End Class

File: Ent_UsAdmin/EmpresasPermitidas.cs
using System;
using System.ComponentModel;
using System.Data.OleDb;
using System.Drawing;
using System.Windows.Forms;


namespace ConfiaAdmin
{

    public partial class EmpresasPermitidas
    {
        private Cargando ncargando;
        public string idUsuario;
        private int ButtonIndex = 0;
        private Cursor DragDropCursor;

        public EmpresasPermitidas()
        {
            InitializeComponent();
        }
        private void EmpresasPermitidas_Load(object sender, EventArgs e)
        {
            CheckForIllegalCrossThreadCalls = false;
            ButtonIndex += 1;

            ncargando = new Cargando();
            ncargando.MonoFlat_Label1.Text = "Cargando Información";
            ncargando.Show();

            FlowEmpresasRestringidas.AllowDrop = true;
            FlowEmpresasPermitidas.AllowDrop = true;
            BackgroundWorker1.RunWorkerAsync();


        }

        private void BackgroundWorker1_DoWork(object sender, DoWorkEventArgs e)
        {
            Module1.iniciarconexionR();
            OleDbCommand comandoEmpresasPermitidas;

            string consultaEmpresasPermitidas;
            OleDbDataReader readerEmpresasPermitidas;
            consultaEmpresasPermitidas = "select id,nombre from empresas where id in (select idempresa from empresaspermitidas where idusuario = '" + idUsuario + "')";
            comandoEmpresasPermitidas = new OleDbCommand();
            comandoEmpresasPermitidas.Connection = Module1.conexionempresaR;
            comandoEmpresasPermitidas.CommandText = consultaEmpresasPermitidas;
            readerEmpresasPermitidas = comandoEmpresasPermitidas.ExecuteReader();
            if (readerEmpresasPermitidas.HasRows)
            {

                while (readerEmpresasPermitidas.Read())
                {
                   Button botonempresa = new Button();

                   // botonempresa.Normalcolor = Color.FromArgb(48, 55, 76);
                  //  botonempresa.Iconimage = null;

                    botonempresa.Size = new Size(197, 48);
                    botonempresa.Name = readerEmpresasPermitidas["id"].ToString();
                    botonempresa.Text = readerEmpresasPermitidas["nombre"].ToString();
                    // botonempresa.Tag = milector("ip")
                    // AddHandler botonempresa.Click, AddressOf clickevenntAsync
                    botonempresa.MouseDown += new MouseEventHandler(OnButtonMouseDown);
                    Invoke(new Action(() => FlowEmpresasPermitidas.Controls.Add(botonempresa)));

                }

            }

            OleDbCommand comandoEmpresasRestringidas;
            string consultaEmpresaRestringidas;
            OleDbDataReader readerEmpresasRestringidas;
            consultaEmpresaRestringidas = "select id,nombre from empresas where id not in (select idempresa from empresaspermitidas where idusuario = '" + idUsuario + "')";
            comandoEmpresasRestringidas = new OleDbCommand();
            comandoEmpresasRestringidas.Connection = Module1.conexionempresaR;
            comandoEmpresasRestringidas.CommandText = consultaEmpresaRestringidas;
            readerEmpresasRestringidas = comandoEmpresasRestringidas.ExecuteReader();
            if (readerEmpresasRestringidas.HasRows)
            {

                while (readerEmpresasRestringidas.Read())
                {
                    Button botonempresa = new Button();

                   // botonempresa.Normalcolor = Color.FromArgb(48, 55, 76);
                    //botonempresa.Iconimage = null;

                    botonempresa.Size = new Size(197, 48);
                    botonempresa.Name = readerEmpresasRestringidas["id"].ToString();
                    botonempresa.Text = readerEmpresasRestringidas["nombre"].ToString();
                    // botonempresa.Tag = milector("ip")
                    // AddHandler botonempresa.Click, AddressOf clickevenntAsync
                    botonempresa.MouseDown += new MouseEventHandler(OnButtonMouseDown);
                    Invoke(new Action(() => FlowEmpresasRestringidas.Controls.Add(botonempresa)));

                }

            }




        }

        private void OnButtonMouseDown(object sender, MouseEventArgs e)
        {

            Button button = (Button)sender;

            using (var bmp = new Bitmap(button.Width, button.Height))
            {
                button.DrawToBitmap(bmp, new Rectangle(Point.Empty, button.Size));
                DragDropCursor = new Cursor(bmp.GetHicon());
            }

            button.Parent.DoDragDrop(button, DragDropEffects.Move);

        }

        private void OnFlowLayoutPanelDragEnter(object sender, DragEventArgs e)
        {

            if (e.AllowedEffect == DragDropEffects.Move && e.Data.GetDataPresent(typeof(Button)))
            {
                e.Effect = DragDropEffects.Move;
            }

        }

        private void OnFlowLayoutPanelDragDrop(object sender, DragEventArgs e)
        {

            Button button = (Button)e.Data.GetData(typeof(Button));
            FlowLayoutPanel destPanel = (FlowLayoutPanel)sender;
            destPanel.Controls.Add(button);
            var targetLoc = destPanel.GetChildAtPoint(destPanel.PointToClient(new Point(e.X, e.Y)));

            if (targetLoc != null)
            {
                int idx = destPanel.Controls.GetChildIndex(targetLoc);
                destPanel.Controls.SetChildIndex(button, idx);
            }

        }

        private void OnFlowLayoutPanelGiveFeedback(object sender, GiveFeedbackEventArgs e)
        {

            e.UseDefaultCursors = false;
            Cursor.Current = DragDropCursor;

        }

        private void EmpresasPermitidas_FormClosing(object sender, FormClosingEventArgs e)
        {
            int num_controles = FlowEmpresasPermitidas.Controls.Count - 1;
            for (int n = num_controles; n >= 0; n -= 1)
            {
                var ctrl = FlowEmpresasPermitidas.Controls[n];
                FlowEmpresasPermitidas.Controls.Remove(ctrl);
                ctrl.Dispose();
            }

            int num_controles2 = FlowEmpresasRestringidas.Controls.Count - 1;
            for (int n = num_controles2; n >= 0; n -= 1)
            {
                var ctrl = FlowEmpresasRestringidas.Controls[n];
                FlowEmpresasRestringidas.Controls.Remove(ctrl);
                ctrl.Dispose();
            }

        }

        private void BackgroundAplicar_DoWork(object sender, DoWorkEventArgs e)
        {
            Module1.iniciarconexionR();

            foreach (Control ctrl in FlowEmpresasPermitidas.Controls)
            {
                if (ctrl is Button)
                {
                    Button btnctrl = (Button)ctrl;
                    OleDbCommand comandoEmpresaPermitida;
                    string consultaEmpresaPermitida;
                    consultaEmpresaPermitida = "if not exists(select id from empresaspermitidas where idempresa = '" + btnctrl.Name + "' and idusuario = '" + idUsuario + @"')
                                            begin
                                              insert into empresasPermitidas values('" + btnctrl.Name + "','" + idUsuario + @"')
                                               end";
                    comandoEmpresaPermitida = new OleDbCommand();
                    comandoEmpresaPermitida.Connection = Module1.conexionempresaR;
                    comandoEmpresaPermitida.CommandText = consultaEmpresaPermitida;
                    comandoEmpresaPermitida.ExecuteNonQuery();



                }
            }


            foreach (Control ctrl in FlowEmpresasRestringidas.Controls)
            {
                if (ctrl is Button)
                {
                 Button btnctrl = (Button)ctrl;
                    OleDbCommand comandoEmpresaRestringida;
                    string consultaEmpresaRestringida;
                    consultaEmpresaRestringida = "if exists(select id from empresaspermitidas where idempresa = '" + btnctrl.Name + "' and idusuario = '" + idUsuario + @"')
                                             begin
                                            delete from empresasPermitidas where idempresa = '" + btnctrl.Name + "' and idusuario = '" + idUsuario + @"' 
                                            end";
                    comandoEmpresaRestringida = new OleDbCommand();
                    comandoEmpresaRestringida.Connection = Module1.conexionempresaR;
                    comandoEmpresaRestringida.CommandText = consultaEmpresaRestringida;
                    comandoEmpresaRestringida.ExecuteNonQuery();



                }
            }

        }

        private void BackgroundWorker1_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            ncargando.Close();

        }

        private void BackgroundAplicar_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            ncargando.Close();
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            ncargando = new Cargando();
            ncargando.MonoFlat_Label1.Text = "Aplicando cambios";
            ncargando.Show();

            BackgroundAplicar.RunWorkerAsync();

        }

        private void Panel1_Paint(object sender, PaintEventArgs e)
        {

        }

        private void Panel1_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                Module_resize.MoveForm(this);
            }
        }
    }
}